### PR TITLE
RAC-989: Revamp user filter (back)

### PR DESCRIPTION
**Pull request by pierrejolly**
**Description (for Contributor and Core Developer)**

**Definition Of Done (for Core Developer only)**

- [x] Tests

### CODE REVIEW

File: src/Akeneo/Platform/Job/back/Domain/Query/FindJobUsersInterface.php [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+ * @author Pierre Jolly <pierre.jolly@akeneo.com>
+ * @copyright 2021 Akeneo SAS (https://www.akeneo.com)
+ * @license https://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
+ */
+interface FindJobUsersInterface
+{
+    /**
+     * @return string[]
+     */
+    public function visible(int $page): array;
```
Comment (StevenVAIDIE) [inline src/Akeneo/Platform/Job/back/Domain/Query/FindJobUsersInterface.php:15]:
```suggestion
interface SearchJobUsersInterface
{
    /**
     * @return string[]
     */
    public function search(int $page): array;
```

File: src/Akeneo/Platform/Job/back/Infrastructure/Controller/GetJobUsersAction.php [hunk 2]
```diff
@@ -0,0 +1,37 @@
+<?php
+
```
Comment (StevenVAIDIE) [inline src/Akeneo/Platform/Job/back/Infrastructure/Controller/GetJobUsersAction.php:2]:
```suggestion
declare(strict_types=1);
```

File: src/Akeneo/Platform/Job/back/Infrastructure/Controller/GetJobUsersAction.php [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        $this->findJobUsers = $findJobUsers;
+    }
+
+    public function __invoke(Request $request): Response
+    {
+        if (!$request->isXmlHttpRequest()) {
+            return new RedirectResponse('/');
+        }
+
+        $page = (int) $request->get('page', 1);
```
Comment (StevenVAIDIE) [inline src/Akeneo/Platform/Job/back/Infrastructure/Controller/GetJobUsersAction.php:31]: Maybe we could create a ticket to add search in this dropdown as we paginate the result

File: src/Akeneo/Platform/Job/back/Infrastructure/Controller/GetJobUsersAction.php [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+        $jobUsers = $this->findJobUsers->visible($page);
```
Comment (StevenVAIDIE) [inline src/Akeneo/Platform/Job/back/Infrastructure/Controller/GetJobUsersAction.php:33]: I don't know if we need to add an Application layer

**Review by StevenVAIDIE — APPROVED** [review]

File: src/Akeneo/Platform/Job/back/Infrastructure/Query/FindJobUsers.php [hunk 5]
```diff
@@ -0,0 +1,56 @@
+<?php
+
```
Comment (StevenVAIDIE) [inline src/Akeneo/Platform/Job/back/Infrastructure/Query/FindJobUsers.php:2]: declare(strict_types=1);

**Review by ghoussard — APPROVED** [review]
Filter appliance and acl (View jobs of all users) handle are missing in this controller : https://github.com/akeneo/pim-community-dev/blob/RAC-process-tracker-revamp/src/Akeneo/Platform/Job/back/Infrastructure/Controller/IndexAction.php

File: src/Akeneo/Platform/Job/back/Infrastructure/Query/FindJobUsers.php (hunk 5 above)
Comment (pierrejolly) [reply src/Akeneo/Platform/Job/back/Infrastructure/Query/FindJobUsers.php:2]: Nice catch, it's missing in `FindJobTypes` too 👍

File: src/Akeneo/Platform/Job/back/Infrastructure/Controller/GetJobUsersAction.php [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+        $this->findJobUsers = $findJobUsers;
+    }
+
+    public function __invoke(Request $request): Response
+    {
+        if (!$request->isXmlHttpRequest()) {
+            return new RedirectResponse('/');
+        }
+
+        if (!$this->securityFacade->isGranted('pim_enrich_job_tracker_view_all_jobs')) {
```
Comment (ghoussard) [inline src/Akeneo/Platform/Job/back/Infrastructure/Controller/GetJobUsersAction.php:42]: I don't know if these lines are very useful. We don't show the filter in frontend if pim_enrich_job_tracker_view_all_jobs is not granted, indeed this route will never be called.

File: src/Akeneo/Platform/Job/back/Infrastructure/Controller/GetJobUsersAction.php [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+        $this->securityFacade = $securityFacade;
+        $this->findJobUsers = $findJobUsers;
+    }
+
+    public function __invoke(Request $request): Response
+    {
+        if (!$request->isXmlHttpRequest()) {
+            return new RedirectResponse('/');
+        }
+
```
Comment (ghoussard) [inline src/Akeneo/Platform/Job/back/Infrastructure/Controller/GetJobUsersAction.php:41]: Is there an acl to view all users or a list of all users ? If yes, we should check it here

**Review by ghoussard — APPROVED** [review]

File: src/Akeneo/Platform/Job/back/Infrastructure/Controller/GetJobUsersAction.php (hunk 7 above)
Comment (StevenVAIDIE) [inline src/Akeneo/Platform/Job/back/Infrastructure/Controller/GetJobUsersAction.php:41]:
```suggestion
        if (!$this->securityFacade->isGranted('pim_enrich_job_tracker_index')) {
            throw new AccessDeniedHttpException('Access forbidden. You are not allowed to list job users.');
        }
```

**Review by StevenVAIDIE — APPROVED** [review]

File: src/Akeneo/Platform/Job/back/Infrastructure/Controller/GetJobUsersAction.php (hunk 6 above)
Comment (pierrejolly) [reply src/Akeneo/Platform/Job/back/Infrastructure/Controller/GetJobUsersAction.php:42]: yes but the route still accessible

File: src/Akeneo/Platform/Job/back/Infrastructure/Controller/GetJobUsersAction.php (hunk 7 above)
Comment (pierrejolly) [reply src/Akeneo/Platform/Job/back/Infrastructure/Controller/GetJobUsersAction.php:41]: Indeed, it's done now ;)

**Pull request merged**